File: agent/sdk/typescript/src/native-capabilities.ts
export const MAX_NATIVE_CAPABILITIES = 238;
export const MAX_NATIVE_BALANCE_VIEWS = 32;
export const MAX_NATIVE_CAPABILITY_BYTES = 65_452;

export type NativeCapability =
  | Readonly<{ kind: "storage_read" | "storage_write" | "emit_event" | "shared_storage_read" | "shared_storage_write" }>
  | Readonly<{ kind: "call"; program: Uint8Array }>
  | Readonly<{ kind: "transfer402"; asset: Uint8Array; to: Uint8Array; maximumAmount: bigint }>
  | Readonly<{ kind: "program_spend"; ownerProgram: Uint8Array; seed: Uint8Array; sourceAccount: Uint8Array; asset: Uint8Array; to: Uint8Array; maximumAmount: bigint }>
  | Readonly<{ kind: "receipt_read"; receiptDigest: Uint8Array }>
  | Readonly<{ kind: "balance_view"; account: Uint8Array; asset: Uint8Array; receiptDigest: Uint8Array }>;

function refuse(): never { throw new TypeError("invalid native capability set"); }
function fixed(value: Uint8Array): Uint8Array {
  if (!(value instanceof Uint8Array) || value.length !== 32 || !value.some((byte) => byte !== 0)) refuse();
  return value;
}
function word(value: bigint, size: number): Uint8Array {
  if (typeof value !== "bigint" || value < 0n || value >= 1n << BigInt(size * 8)) refuse();
  const bytes = new Uint8Array(size);
  for (let index = size - 1; index >= 0; index--) { bytes[index] = Number(value & 255n); value >>= 8n; }
  return bytes;
}
function join(...parts: readonly Uint8Array[]): Uint8Array {
  const output = new Uint8Array(parts.reduce((size, part) => size + part.length, 0));
  let offset = 0; for (const part of parts) { output.set(part, offset); offset += part.length; }
  return output;
}
function compare(left: Uint8Array, right: Uint8Array): number {
  for (let index = 0; index < Math.min(left.length, right.length); index++) {
    const delta = left[index]! - right[index]!; if (delta !== 0) return delta;
  }
  return left.length - right.length;
}
function key(grant: NativeCapability): readonly Uint8Array[] {
  const rank = (value: number) => Uint8Array.of(value);
  switch (grant.kind) {
    case "storage_read": return [rank(0)];
    case "storage_write": return [rank(1)];
    case "emit_event": return [rank(2)];
    case "call": return [rank(3), grant.program];
    case "transfer402": return [rank(4), grant.asset, grant.to];
    case "program_spend": return [rank(5), grant.ownerProgram, grant.seed, grant.sourceAccount, grant.asset, grant.to];
    case "receipt_read": return [rank(6), grant.receiptDigest];
    case "balance_view": return [rank(7), grant.account, grant.asset];
    case "shared_storage_read": return [rank(8)];
    case "shared_storage_write": return [rank(9)];
    default: return refuse();
  }
}
function compareGrants(left: NativeCapability, right: NativeCapability): number {
  const leftKey = key(left), rightKey = key(right);
  for (let index = 0; index < Math.min(leftKey.length, rightKey.length); index++) {
    const delta = compare(leftKey[index]!, rightKey[index]!); if (delta !== 0) return delta;
  }
  return leftKey.length - rightKey.length;
}
function snapshot(grant: NativeCapability): NativeCapability {
  return Object.freeze(Object.fromEntries(Object.entries(grant).map(([name, value]) => [name, value instanceof Uint8Array ? new Uint8Array(value) : value]))) as NativeCapability;
}

export async function deriveNativeProgramAccount(owner: Uint8Array, seed: Uint8Array): Promise<Uint8Array> {
  fixed(owner); if (!(seed instanceof Uint8Array) || seed.length > 128) refuse();
  const bytes = join(new TextEncoder().encode("LayerX/programs/program-account/v1\0"), owner, word(BigInt(seed.length), 4), seed);
  return new Uint8Array(await crypto.subtle.digest("SHA-256", new Uint8Array(bytes)));
}

async function encodeGrant(grant: NativeCapability): Promise<Uint8Array> {
  switch (grant.kind) {
    case "storage_read": return Uint8Array.of(1);
    case "storage_write": return Uint8Array.of(2);
    case "emit_event": return Uint8Array.of(3);
    case "call": return join(Uint8Array.of(4), fixed(grant.program));
    case "transfer402":
      if (grant.maximumAmount === 0n) refuse();
      return join(Uint8Array.of(5), fixed(grant.asset), fixed(grant.to), word(grant.maximumAmount, 16));
    case "program_spend": {
      if (grant.maximumAmount === 0n || compare(await deriveNativeProgramAccount(grant.ownerProgram, grant.seed), grant.sourceAccount) !== 0) refuse();
      return join(Uint8Array.of(9), fixed(grant.ownerProgram), word(BigInt(grant.seed.length), 2), grant.seed,
        fixed(grant.sourceAccount), fixed(grant.asset), fixed(grant.to), word(grant.maximumAmount, 16));
    }
    case "receipt_read": return join(Uint8Array.of(6), fixed(grant.receiptDigest));
    case "balance_view": return join(Uint8Array.of(10), fixed(grant.account), fixed(grant.asset), fixed(grant.receiptDigest));
    case "shared_storage_read": return Uint8Array.of(7);
    case "shared_storage_write": return Uint8Array.of(8);
    default: return refuse();
  }
}

export async function encodeNativeCapabilitySet(grants: readonly NativeCapability[]): Promise<Uint8Array> {
  if (grants.length > MAX_NATIVE_CAPABILITIES) refuse();
  const ordered = grants.map(snapshot).sort(compareGrants);
  if (ordered.filter((grant) => grant.kind === "balance_view").length > MAX_NATIVE_BALANCE_VIEWS) refuse();
  for (let index = 1; index < ordered.length; index++) if (compareGrants(ordered[index - 1]!, ordered[index]!) === 0) refuse();
  const encoded = join(word(BigInt(ordered.length), 2), ...await Promise.all(ordered.map(encodeGrant)));
  if (encoded.length > MAX_NATIVE_CAPABILITY_BYTES) refuse();
  return encoded;
}

class Cursor {
  private offset = 0;
  constructor(private readonly bytes: Uint8Array) {}
  take(length: number): Uint8Array {
    if (length < 0 || length > this.bytes.length - this.offset) refuse();
    const value = this.bytes.slice(this.offset, this.offset + length); this.offset += length; return value;
  }
  number(length: number): bigint { let value = 0n; for (const byte of this.take(length)) value = (value << 8n) | BigInt(byte); return value; }
  end(): void { if (this.offset !== this.bytes.length) refuse(); }
}
function decodeGrant(reader: Cursor): NativeCapability {
  switch (Number(reader.number(1))) {
    case 1: return { kind: "storage_read" };
    case 2: return { kind: "storage_write" };
    case 3: return { kind: "emit_event" };
    case 4: return { kind: "call", program: reader.take(32) };
    case 5: return { kind: "transfer402", asset: reader.take(32), to: reader.take(32), maximumAmount: reader.number(16) };
    case 6: return { kind: "receipt_read", receiptDigest: reader.take(32) };
    case 7: return { kind: "shared_storage_read" };
    case 8: return { kind: "shared_storage_write" };
    case 9: {
      const ownerProgram = reader.take(32), length = Number(reader.number(2)); if (length > 128) refuse();
      return { kind: "program_spend", ownerProgram, seed: reader.take(length), sourceAccount: reader.take(32), asset: reader.take(32), to: reader.take(32), maximumAmount: reader.number(16) };
    }
    case 10: return { kind: "balance_view", account: reader.take(32), asset: reader.take(32), receiptDigest: reader.take(32) };
    default: return refuse();
  }
}
export async function decodeNativeCapabilitySet(bytes: Uint8Array): Promise<readonly NativeCapability[]> {
  if (!(bytes instanceof Uint8Array)) refuse();
  bytes = new Uint8Array(bytes);
  if (bytes.length < 2 || bytes.length > MAX_NATIVE_CAPABILITY_BYTES) refuse();
  const reader = new Cursor(bytes), count = Number(reader.number(2)); if (count > MAX_NATIVE_CAPABILITIES) refuse();
  const grants = Array.from({ length: count }, () => decodeGrant(reader)); reader.end();
  if (compare(await encodeNativeCapabilitySet(grants), bytes) !== 0) refuse();
  return Object.freeze(grants.map(snapshot));
}

export async function narrowNativeCapabilitySet(parent: readonly NativeCapability[], requested: readonly NativeCapability[]): Promise<readonly NativeCapability[]> {
  if (parent.length > MAX_NATIVE_CAPABILITIES || requested.length > MAX_NATIVE_CAPABILITIES) refuse();
  const parentSnapshot = parent.map(snapshot), childSnapshot = requested.map(snapshot);
  const parents = await decodeNativeCapabilitySet(await encodeNativeCapabilitySet(parentSnapshot));
  const children = await decodeNativeCapabilitySet(await encodeNativeCapabilitySet(childSnapshot));
  for (const child of children) {
    const ancestor = parents.find((grant) => compareGrants(grant, child) === 0); if (ancestor === undefined) refuse();
    if ((child.kind === "transfer402" && ancestor.kind === "transfer402" || child.kind === "program_spend" && ancestor.kind === "program_spend") && child.maximumAmount > ancestor.maximumAmount) refuse();
    if (child.kind === "balance_view" && ancestor.kind === "balance_view" && compare(child.receiptDigest, ancestor.receiptDigest) !== 0) refuse();
  }
  return children;
}

File: agent/crates/layerx-sdk/src/native_capabilities.rs
use sha2::{Digest as _, Sha256};

pub const MAX_NATIVE_CAPABILITIES: usize = 238;
pub const MAX_NATIVE_BALANCE_VIEWS: usize = 32;
pub const MAX_NATIVE_CAPABILITY_BYTES: usize = 65_452;

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum NativeCapability {
    StorageRead,
    StorageWrite,
    EmitEvent,
    Call {
        program: [u8; 32],
    },
    Transfer402 {
        asset: [u8; 32],
        to: [u8; 32],
        maximum_amount: u128,
    },
    ProgramSpend {
        owner_program: [u8; 32],
        seed: Vec<u8>,
        source_account: [u8; 32],
        asset: [u8; 32],
        to: [u8; 32],
        maximum_amount: u128,
    },
    ReceiptRead {
        receipt_digest: [u8; 32],
    },
    BalanceView {
        account: [u8; 32],
        asset: [u8; 32],
        receipt_digest: [u8; 32],
    },
    SharedStorageRead,
    SharedStorageWrite,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum NativeCapabilityError {
    Bounds,
    InvalidGrant,
    Duplicate,
    NonCanonical,
    Widening,
}

impl NativeCapability {
    fn key(&self) -> (u8, Vec<Vec<u8>>) {
        match self {
            Self::StorageRead => (0, vec![]),
            Self::StorageWrite => (1, vec![]),
            Self::EmitEvent => (2, vec![]),
            Self::Call { program } => (3, vec![program.to_vec()]),
            Self::Transfer402 { asset, to, .. } => (4, vec![asset.to_vec(), to.to_vec()]),
            Self::ProgramSpend {
                owner_program,
                seed,
                source_account,
                asset,
                to,
                ..
            } => (
                5,
                vec![
                    owner_program.to_vec(),
                    seed.clone(),
                    source_account.to_vec(),
                    asset.to_vec(),
                    to.to_vec(),
                ],
            ),
            Self::ReceiptRead { receipt_digest } => (6, vec![receipt_digest.to_vec()]),
            Self::BalanceView { account, asset, .. } => (7, vec![account.to_vec(), asset.to_vec()]),
            Self::SharedStorageRead => (8, vec![]),
            Self::SharedStorageWrite => (9, vec![]),
        }
    }

    fn validate(&self) -> Result<(), NativeCapabilityError> {
        let valid = match self {
            Self::Call { program } => *program != [0; 32],
            Self::Transfer402 {
                asset,
                to,
                maximum_amount,
            } => *asset != [0; 32] && *to != [0; 32] && *maximum_amount != 0,
            Self::ProgramSpend {
                owner_program,
                seed,
                source_account,
                asset,
                to,
                maximum_amount,
            } => {
                *asset != [0; 32]
                    && *to != [0; 32]
                    && *maximum_amount != 0
                    && derive_native_program_account(*owner_program, seed)? == *source_account
            }
            Self::ReceiptRead { receipt_digest } => *receipt_digest != [0; 32],
            Self::BalanceView {
                account,
                asset,
                receipt_digest,
            } => *account != [0; 32] && *asset != [0; 32] && *receipt_digest != [0; 32],
            _ => true,
        };
        if valid {
            Ok(())
        } else {
            Err(NativeCapabilityError::InvalidGrant)
        }
    }

    fn append(&self, output: &mut Vec<u8>) -> Result<(), NativeCapabilityError> {
        match self {
            Self::StorageRead => output.push(1),
            Self::StorageWrite => output.push(2),
            Self::EmitEvent => output.push(3),
            Self::Call { program } => {
                output.push(4);
                output.extend_from_slice(program);
            }
            Self::Transfer402 {
                asset,
                to,
                maximum_amount,
            } => {
                output.push(5);
                output.extend_from_slice(asset);
                output.extend_from_slice(to);
                output.extend_from_slice(&maximum_amount.to_be_bytes());
            }
            Self::ProgramSpend {
                owner_program,
                seed,
                source_account,
                asset,
                to,
                maximum_amount,
            } => {
                output.push(9);
                output.extend_from_slice(owner_program);
                output.extend_from_slice(
                    &u16::try_from(seed.len())
                        .map_err(|_| NativeCapabilityError::Bounds)?
                        .to_be_bytes(),
                );
                output.extend_from_slice(seed);
                output.extend_from_slice(source_account);
                output.extend_from_slice(asset);
                output.extend_from_slice(to);
                output.extend_from_slice(&maximum_amount.to_be_bytes());
            }
            Self::ReceiptRead { receipt_digest } => {
                output.push(6);
                output.extend_from_slice(receipt_digest);
            }
            Self::BalanceView {
                account,
                asset,
                receipt_digest,
            } => {
                output.push(10);
                output.extend_from_slice(account);
                output.extend_from_slice(asset);
                output.extend_from_slice(receipt_digest);
            }
            Self::SharedStorageRead => output.push(7),
            Self::SharedStorageWrite => output.push(8),
        }
        Ok(())
    }
}

/// # Errors
/// Refuses a zero owner or a seed exceeding the frozen 128-byte bound.
pub fn derive_native_program_account(
    owner: [u8; 32],
    seed: &[u8],
) -> Result<[u8; 32], NativeCapabilityError> {
    if owner == [0; 32] || seed.len() > 128 {
        return Err(NativeCapabilityError::InvalidGrant);
    }
    let mut hash = Sha256::new();
    hash.update(b"LayerX/programs/program-account/v1\0");
    hash.update(owner);
    hash.update(
        u32::try_from(seed.len())
            .map_err(|_| NativeCapabilityError::Bounds)?
            .to_be_bytes(),
    );
    hash.update(seed);
    Ok(hash.finalize().into())
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct NativeCapabilitySet(Vec<NativeCapability>);

impl NativeCapabilitySet {
    /// # Errors
    /// Refuses invalid grants, duplicate authority keys, and frozen count bounds.
    pub fn new(mut grants: Vec<NativeCapability>) -> Result<Self, NativeCapabilityError> {
        if grants.len() > MAX_NATIVE_CAPABILITIES
            || grants
                .iter()
                .filter(|grant| matches!(grant, NativeCapability::BalanceView { .. }))
                .count()
                > MAX_NATIVE_BALANCE_VIEWS
        {
            return Err(NativeCapabilityError::Bounds);
        }
        for grant in &grants {
            grant.validate()?;
        }
        grants.sort_by_key(NativeCapability::key);
        if grants.windows(2).any(|pair| pair[0].key() == pair[1].key()) {
            return Err(NativeCapabilityError::Duplicate);
        }
        Ok(Self(grants))
    }

    #[must_use]
    pub fn grants(&self) -> &[NativeCapability] {
        &self.0
    }

    /// # Errors
    /// Refuses any encoding exceeding the frozen canonical byte bound.
    pub fn encode(&self) -> Result<Vec<u8>, NativeCapabilityError> {
        let mut output = u16::try_from(self.0.len())
            .map_err(|_| NativeCapabilityError::Bounds)?
            .to_be_bytes()
            .to_vec();
        for grant in &self.0 {
            grant.append(&mut output)?;
        }
        if output.len() > MAX_NATIVE_CAPABILITY_BYTES {
            return Err(NativeCapabilityError::Bounds);
        }
        Ok(output)
    }

    /// # Errors
    /// Refuses unknown tags, malformed fields, duplicate keys, and noncanonical ordering.
    pub fn decode(encoded: &[u8]) -> Result<Self, NativeCapabilityError> {
        if encoded.len() < 2 || encoded.len() > MAX_NATIVE_CAPABILITY_BYTES {
            return Err(NativeCapabilityError::Bounds);
        }
        let mut remaining = encoded;
        let count = usize::from(u16::from_be_bytes(take(&mut remaining)?));
        if count > MAX_NATIVE_CAPABILITIES {
            return Err(NativeCapabilityError::Bounds);
        }
        let mut grants = Vec::with_capacity(count);
        for _ in 0..count {
            grants.push(decode_grant(&mut remaining)?);
        }
        let result = Self::new(grants)?;
        if !remaining.is_empty() || result.encode()? != encoded {
            return Err(NativeCapabilityError::NonCanonical);
        }
        Ok(result)
    }

    /// # Errors
    /// Refuses new authority keys, increased amounts, or changed `BalanceView` receipt digests.
    pub fn narrow(&self, requested: Self) -> Result<Self, NativeCapabilityError> {
        for child in &requested.0 {
            let parent = self
                .0
                .iter()
                .find(|parent| parent.key() == child.key())
                .ok_or(NativeCapabilityError::Widening)?;
            let admitted = match (parent, child) {
                (
                    NativeCapability::Transfer402 {
                        maximum_amount: ceiling,
                        ..
                    },
                    NativeCapability::Transfer402 { maximum_amount, .. },
                )
                | (
                    NativeCapability::ProgramSpend {
                        maximum_amount: ceiling,
                        ..
                    },
                    NativeCapability::ProgramSpend { maximum_amount, .. },
                ) => maximum_amount <= ceiling,
                (
                    NativeCapability::BalanceView {
                        receipt_digest: expected,
                        ..
                    },
                    NativeCapability::BalanceView { receipt_digest, .. },
                ) => receipt_digest == expected,
                _ => true,
            };
            if !admitted {
                return Err(NativeCapabilityError::Widening);
            }
        }
        Ok(requested)
    }
}

fn take<const LENGTH: usize>(remaining: &mut &[u8]) -> Result<[u8; LENGTH], NativeCapabilityError> {
    let bytes = remaining
        .get(..LENGTH)
        .ok_or(NativeCapabilityError::NonCanonical)?;
    let result = bytes
        .try_into()
        .map_err(|_| NativeCapabilityError::NonCanonical)?;
    *remaining = &remaining[LENGTH..];
    Ok(result)
}

fn decode_grant(remaining: &mut &[u8]) -> Result<NativeCapability, NativeCapabilityError> {
    Ok(match take::<1>(remaining)?[0] {
        1 => NativeCapability::StorageRead,
        2 => NativeCapability::StorageWrite,
        3 => NativeCapability::EmitEvent,
        4 => NativeCapability::Call {
            program: take(remaining)?,
        },
        5 => NativeCapability::Transfer402 {
            asset: take(remaining)?,
            to: take(remaining)?,
            maximum_amount: u128::from_be_bytes(take(remaining)?),
        },
        6 => NativeCapability::ReceiptRead {
            receipt_digest: take(remaining)?,
        },
        7 => NativeCapability::SharedStorageRead,
        8 => NativeCapability::SharedStorageWrite,
        9 => {
            let owner_program = take(remaining)?;
            let length = usize::from(u16::from_be_bytes(take(remaining)?));
            if length > 128 {
                return Err(NativeCapabilityError::Bounds);
            }
            let seed = remaining
                .get(..length)
                .ok_or(NativeCapabilityError::NonCanonical)?
                .to_vec();
            *remaining = &remaining[length..];
            NativeCapability::ProgramSpend {
                owner_program,
                seed,
                source_account: take(remaining)?,
                asset: take(remaining)?,
                to: take(remaining)?,
                maximum_amount: u128::from_be_bytes(take(remaining)?),
            }
        }
        10 => NativeCapability::BalanceView {
            account: take(remaining)?,
            asset: take(remaining)?,
            receipt_digest: take(remaining)?,
        },
        _ => return Err(NativeCapabilityError::NonCanonical),
    })
}
